module_2/gifts/repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -0,0 +1,111 @@
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.mapper.CertificateRowMapper;
+import com.epam.esm.repository.model.CertificateModel;
+import com.epam.esm.repository.model.TagModel;
+
+@Repository
+public class CertificateRepositoryImpl implements CertificateRepository {
+	private static final String INSERT_QUERY = "INSERT INTO gift_certificates "
+			+ "(name, description, price, duration, create_date, last_update_date) VALUES (?, ?, ?, ?, ?, ?)";
+	private static final String SELECT_CERTIFICATE_BY_ID_QUERY = "SELECT id, name, description, price, duration, create_date, last_update_date, is_deleted FROM gift_certificates WHERE id = ? AND is_deleted = 0";
+	private static final String SELECT_CERTIFICATE_BY_NAME_QUERY = "SELECT id, name, description, price, duration, create_date, last_update_date, is_deleted FROM gift_certificates WHERE name = ?";
+	private static final String SELECT_CERTIFICATE_BY_TAG_ID_QUERY = "SELECT id, name, description, price, duration, create_date, last_update_date, is_deleted FROM gift_certificates INNER JOIN tags_certificates "
+			+ "ON id = certificate_id WHERE tag_id = ?";
+	private static final String REMOVE_CERTIFICATE__QUERY = "UPDATE gift_certificates SET is_deleted = true WHERE id = ?";
+	private static final String UPDATE_CERTIFICATE__QUERY = "UPDATE gift_certificates SET name = ?, description = ?, price = ?, duration = ?, last_update_date = ? WHERE id = ?";
+
+	@Autowired
+	private JdbcTemplate jdbcTemplate;
+
+	@Autowired
+	private CertificateRowMapper certificateRowMapper;
+
+	@Override
+	public CertificateModel create(CertificateModel certificateModel) {
+		KeyHolder keyHolder = new GeneratedKeyHolder();
+
+		jdbcTemplate.update(connection -> {
+			PreparedStatement preparedStatement = connection.prepareStatement(INSERT_QUERY,
+					PreparedStatement.RETURN_GENERATED_KEYS);
+			preparedStatement.setString(1, certificateModel.getName());
+			preparedStatement.setString(2, certificateModel.getDescription());
+			preparedStatement.setBigDecimal(3, certificateModel.getPrice());
+			preparedStatement.setInt(4, certificateModel.getDuration());
+			preparedStatement.setTimestamp(5, Timestamp.valueOf(certificateModel.getCreateDate()));
+			preparedStatement.setTimestamp(6, Timestamp.valueOf(certificateModel.getLastUpdateDate()));
+			return preparedStatement;
+		}, keyHolder);
+
+		return readById(keyHolder.getKey().longValue());
+	}
+
+	@Override
+	public CertificateModel readById(long certificateId) {
+		List<CertificateModel> certificateModelList = jdbcTemplate.query(SELECT_CERTIFICATE_BY_ID_QUERY,
+				certificateRowMapper, certificateId);
+		if (certificateModelList.isEmpty()) {
+			return null;
```

Comment:
@LizavetaSinitsyna It's not good practice to return `null`. We can throw exception if it not found.

Reply:
If we throw exception directly in DAO, Controller layer will know about DAO. As I understand the task of DAO layer is to return what it found, but not evaluate whether the return value is correct or not. Evaluation of DAO layer results is something that service layer should make. Please correct me if I'm not right. 
In addition, all keys for exception messages in .properties file are hold in ErrorCode.class in Service layer, so shoud we create a different logic for exception and hold them both in service layer and dao layer at the same time? If we move ErrorCode.class to dao layer it seems to be not logicaly correct because this class contains mostly service error codes. Of cource, we can assign ErrorCode value directly on controller layer in exception handler class, but is it a good practice to handle this exception on Controller layer? I thought that they shouldn't know about each other, but maybe I somehow misunderstand the logic of layer division.

Reply:
@LizavetaSinitsyna It can be Repository exception which is caught in the handler.

Reply:
a good practice to return Optional and if needed to throw exception on service level